File: snowpipe-streaming-python/src/snowpipe_streaming_manager.py
import logging
from typing import List, Dict, Any, Optional
from snowflake.ingest.streaming import StreamingIngestClient, StreamingIngestChannel
from snowflake.ingest.streaming.streaming_ingest_error import StreamingIngestError
from models import Order, OrderItem
from config_manager import ConfigManager
import snowflake.connector
from cryptography.hazmat.primitives import serialization
from cryptography.hazmat.backends import default_backend
import random
import time

logger = logging.getLogger(__name__)


class SnowpipeStreamingManager:
    def __init__(self, config: ConfigManager, instance_id: int = -1):
        self.config = config
        self.instance_id = instance_id
        self._last_orders_offset: str | None = None
        self._last_order_items_offset: str | None = None
        
        channel_suffix = f"_instance_{instance_id}" if instance_id >= 0 else ""
        logger.info(
            f"Creating Snowflake Streaming clients and opening channels"
            f"{' for instance ' + str(instance_id) if instance_id >= 0 else '...'}"
        )
        
        self.properties = {
            "account": config.get_snowflake_account(),
            "user": config.get_snowflake_user(),
            "private_key": config.get_private_key(),
            "url": config.get_snowflake_url(),
            "role": config.get_role(),
            "warehouse": config.get_warehouse(),
        }
        
        self.orders_client = StreamingIngestClient(
            client_name=f"ORDERS_CLIENT_{instance_id}",
            db_name=config.get_database(),
            schema_name=config.get_schema(),
            pipe_name=config.get_property("pipe.orders.name"),
            properties=self.properties,
        )
        
        self.order_items_client = StreamingIngestClient(
            client_name=f"ORDER_ITEMS_CLIENT_{instance_id}",
            db_name=config.get_database(),
            schema_name=config.get_schema(),
            pipe_name=config.get_property("pipe.order_items.name"),
            properties=self.properties,
        )
        
        self.orders_channel = self._open_channel(
            self.orders_client,
            config.get_property("channel.orders.name") + channel_suffix,
        )
        
        self.order_items_channel = self._open_channel(
            self.order_items_client,
            config.get_property("channel.order_items.name") + channel_suffix,
        )
        
        logger.info("All clients and channels initialized successfully")

    def _open_channel(
        self, client: StreamingIngestClient, channel_name: str
    ) -> StreamingIngestChannel:
        initial_offset = "0"
        channel, status = client.open_channel(channel_name, initial_offset)
        
        latest_offset = channel.get_latest_committed_offset_token()
        logger.info(
            f"Channel {channel_name} opened. Latest committed offset: "
            f"{latest_offset if latest_offset else 'NULL (new channel)'}"
        )
        
        return channel

    def get_max_customer_id(self) -> int:
        private_key_pem = self.config.get_private_key()
        private_key_obj = serialization.load_pem_private_key(
            private_key_pem.encode(),
            password=None,
            backend=default_backend()
        )
        
        conn_params = {
            "account": self.config.get_snowflake_account(),
            "user": self.config.get_snowflake_user(),
            "role": self.config.get_role(),
            "warehouse": self.config.get_warehouse(),
            "database": self.config.get_database(),
            "schema": self.config.get_schema(),
            "private_key": private_key_obj,
        }
        
        max_id = 1
        try:
            conn = snowflake.connector.connect(**conn_params)
            cursor = conn.cursor()
            cursor.execute(
                f"SELECT MAX(CUSTOMER_ID) as MAX_ID FROM "
                f"{self.config.get_database()}.RAW.CUSTOMERS"
            )
            result = cursor.fetchone()
            if result and result[0]:
                max_id = result[0]
            logger.info(f"Max customer ID in database: {max_id}")
            cursor.close()
            conn.close()
        except Exception as e:
            logger.error(f"Error fetching max customer ID: {e}")
            raise
        
        return max_id

    def get_customer_segment(self, customer_id: int) -> str:
        """
        Get customer segment for a given customer_id.
        For new customers (beyond max_customer_id), randomly assign segment.
        For existing customers, query from database.
        """
        import random
        
        try:
            # For simplicity during high-volume streaming, randomly assign segments
            # to avoid frequent DB queries that could slow down ingestion
            segments = ["Premium", "Standard", "Basic"]
            return random.choice(segments)
        except Exception as e:
            logger.warning(f"Error getting customer segment: {e}, defaulting to Standard")
            return "Standard"

    def insert_order(self, order: Order) -> None:
        row = order.to_dict()
        offset_token = f"order_{order.order_id}"
        
        self.orders_channel.append_row(row, offset_token)
        logger.debug(f"Order {order.order_id} inserted with offset {offset_token}")

    def insert_orders(self, orders: List[Order]) -> None:
        if not orders:
            return
        
        start_offset = f"order_{orders[0].order_id}"
        end_offset = f"order_{orders[-1].order_id}"
        
        rows = [order.to_dict() for order in orders]
        
        self._insert_with_backpressure_retry(
            self.orders_channel, rows, start_offset, end_offset, "orders"
        )
        self._last_orders_offset = end_offset
        logger.debug(
            f"Inserted {len(orders)} orders (offset range: {start_offset} to {end_offset})"
        )

    def insert_order_items(self, items: List[OrderItem]) -> None:
        if not items:
            return
        
        start_offset = f"item_{items[0].order_item_id}"
        end_offset = f"item_{items[-1].order_item_id}"
        
        rows = [item.to_dict() for item in items]
        
        self._insert_with_backpressure_retry(
            self.order_items_channel, rows, start_offset, end_offset, "order_items"
        )
        self._last_order_items_offset = end_offset
        logger.debug(
            f"Inserted {len(items)} order items (offset range: {start_offset} to {end_offset})"
        )

    def _insert_with_backpressure_retry(
        self,
        channel: StreamingIngestChannel,
        rows: List[Dict[str, Any]],
        start_offset: str,
        end_offset: str,
        data_type: str,
        max_retries: int = 5,
        initial_delay: float = 1.0,
        max_delay: float = 30.0,
    ) -> None:
        """
        Insert rows with exponential backoff retry for ReceiverSaturated errors.
        
        Args:
            channel: The streaming channel to insert into
            rows: Data rows to insert
            start_offset: Starting offset token
            end_offset: Ending offset token
            data_type: Description of data type (for logging)
            max_retries: Maximum number of retry attempts
            initial_delay: Initial delay in seconds before first retry
            max_delay: Maximum delay between retries
        """
        delay = initial_delay
        
        for attempt in range(max_retries):
            try:
                channel.append_rows(rows, start_offset, end_offset)
                
                if attempt > 0:
                    logger.info(
                        f"Successfully inserted {len(rows)} {data_type} after {attempt + 1} attempts"
                    )
                return
                
            except StreamingIngestError as e:
                error_msg = str(e)
                
                # Check for ReceiverSaturated (HTTP 429) backpressure errors
                if "ReceiverSaturated" in error_msg or "429" in error_msg:
                    if attempt < max_retries - 1:
                        jitter = random.uniform(0, delay * 0.25)
                        logger.warning(
                            f"Backpressure detected for {data_type} (attempt {attempt + 1}/{max_retries}): "
                            f"Channel buffers full. Retrying in {delay + jitter:.1f}s..."
                        )
                        time.sleep(delay + jitter)
                        delay = min(delay * 2, max_delay)  # Exponential backoff
                        continue
                    else:
                        logger.error(
                            f"Failed to insert {len(rows)} {data_type} after {max_retries} attempts: "
                            f"Channel buffers remain saturated"
                        )
                        raise
                else:
                    # Non-backpressure error, re-raise immediately
                    logger.error(f"Unexpected error inserting {data_type}: {error_msg}")
                    raise
            
            except Exception as e:
                logger.error(f"Unexpected error type inserting {data_type}: {type(e).__name__}: {e}")
                raise
    
    def get_latest_order_offset(self) -> Optional[str]:
        return self.orders_channel.get_latest_committed_offset_token()

    def get_latest_order_item_offset(self) -> Optional[str]:
        return self.order_items_channel.get_latest_committed_offset_token()

    def wait_for_flush(self, timeout_seconds: int = 120, poll_interval: float = 2.0) -> bool:
        """
        Wait until both channels have committed all in-flight data.
        Polls each channel's latest committed offset token until it matches
        the last offset we sent, confirming Snowflake has received everything.

        Returns True if both channels flushed within the timeout, False otherwise.
        """
        logger.info("Waiting for all channels to flush in-flight data...")
        start = time.time()

        channels_to_check = []
        if self._last_orders_offset:
            channels_to_check.append(
                ("orders", self.orders_channel, self._last_orders_offset)
            )
        if self._last_order_items_offset:
            channels_to_check.append(
                ("order_items", self.order_items_channel, self._last_order_items_offset)
            )

        if not channels_to_check:
            logger.info("No data was sent — nothing to flush")
            return True

        flushed = {name: False for name, _, _ in channels_to_check}

        while time.time() - start < timeout_seconds:
            all_done = True
            for name, channel, expected_offset in channels_to_check:
                if flushed[name]:
                    continue
                committed = channel.get_latest_committed_offset_token()
                if committed == expected_offset:
                    elapsed = time.time() - start
                    logger.info(
                        f"Channel {name} flushed (offset {committed}) in {elapsed:.1f}s"
                    )
                    flushed[name] = True
                else:
                    all_done = False

            if all_done:
                total = time.time() - start
                logger.info(f"All channels flushed in {total:.1f}s")
                return True

            time.sleep(poll_interval)

        # Timeout — log which channels are still pending
        for name, channel, expected_offset in channels_to_check:
            if not flushed[name]:
                committed = channel.get_latest_committed_offset_token()
                logger.warning(
                    f"Channel {name} NOT flushed after {timeout_seconds}s "
                    f"(expected: {expected_offset}, committed: {committed})"
                )
        return False

    def close(self) -> None:
        try:
            logger.info("Closing channels...")
            if hasattr(self, "orders_channel"):
                self.orders_channel.close()
            if hasattr(self, "order_items_channel"):
                self.order_items_channel.close()
            
            logger.info("Closing clients...")
            if hasattr(self, "orders_client"):
                self.orders_client.close()
            if hasattr(self, "order_items_client"):
                self.order_items_client.close()
            
            logger.info("Snowpipe Streaming manager closed successfully")
        except Exception as e:
            logger.error(f"Error closing Snowpipe Streaming manager: {e}")

File: snowpipe-streaming-java/src/main/java/com/snowflake/demo/SnowpipeStreamingManager.java
package com.snowflake.demo;

import com.snowflake.ingest.streaming.SnowflakeStreamingIngestChannel;
import com.snowflake.ingest.streaming.SnowflakeStreamingIngestClient;
import com.snowflake.ingest.streaming.SnowflakeStreamingIngestClientFactory;
import com.snowflake.ingest.streaming.OpenChannelResult;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.security.KeyFactory;
import java.security.PrivateKey;
import java.security.spec.PKCS8EncodedKeySpec;
import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.ResultSet;
import java.sql.Statement;
import java.util.ArrayList;
import java.util.Base64;
import java.util.List;
import java.util.Map;
import java.util.Properties;
import java.util.stream.Collectors;

public class SnowpipeStreamingManager {
    private static final Logger logger = LoggerFactory.getLogger(SnowpipeStreamingManager.class);

    private final SnowflakeStreamingIngestClient ordersClient;
    private final SnowflakeStreamingIngestClient orderItemsClient;
    private final SnowflakeStreamingIngestChannel ordersChannel;
    private final SnowflakeStreamingIngestChannel orderItemsChannel;
    private final ConfigManager config;
    private final Properties connectionProps;
    private volatile String lastOrdersOffset;
    private volatile String lastOrderItemsOffset;

    public SnowpipeStreamingManager(ConfigManager config) throws Exception {
        this(config, -1);
    }

    public SnowpipeStreamingManager(ConfigManager config, int instanceId) throws Exception {
        this.config = config;

        this.connectionProps = new Properties();
        connectionProps.put("account", config.getSnowflakeAccount());
        connectionProps.put("user", config.getSnowflakeUser());
        connectionProps.put("url", config.getSnowflakeUrl());
        connectionProps.put("private_key", config.getPrivateKey());
        connectionProps.put("role", config.getRole());
        connectionProps.put("warehouse", config.getWarehouse());
        connectionProps.put("db", config.getDatabase());
        connectionProps.put("schema", config.getSchema());

        String channelSuffix = (instanceId >= 0) ? "_instance_" + instanceId : "";
        logger.info("Creating Snowflake Streaming clients and opening channels{}", 
                   instanceId >= 0 ? " for instance " + instanceId : "...");
        
        this.ordersClient = SnowflakeStreamingIngestClientFactory.builder(
                "ORDERS_CLIENT_" + java.util.UUID.randomUUID(),
                config.getDatabase(),
                config.getSchema(),
                config.getProperty("pipe.orders.name")
        )
                .setProperties(connectionProps)
                .build();
        
        this.orderItemsClient = SnowflakeStreamingIngestClientFactory.builder(
                "ORDER_ITEMS_CLIENT_" + java.util.UUID.randomUUID(),
                config.getDatabase(),
                config.getSchema(),
                config.getProperty("pipe.order_items.name")
        )
                .setProperties(connectionProps)
                .build();

        this.ordersChannel = openChannel(
                ordersClient,
                config.getProperty("channel.orders.name") + channelSuffix
        );

        this.orderItemsChannel = openChannel(
                orderItemsClient,
                config.getProperty("channel.order_items.name") + channelSuffix
        );

        logger.info("All clients and channels initialized successfully");
    }

    private SnowflakeStreamingIngestChannel openChannel(SnowflakeStreamingIngestClient client, String channelName) throws Exception {
        String initialOffset = "0";
        OpenChannelResult result = client.openChannel(channelName, initialOffset);
        SnowflakeStreamingIngestChannel channel = result.getChannel();
        
        String latestOffset = channel.getLatestCommittedOffsetToken();
        logger.info("Channel {} opened. Latest committed offset: {}", channelName, 
                    latestOffset != null ? latestOffset : "NULL (new channel)");
        
        return channel;
    }

    private PrivateKey parsePrivateKey(String privateKeyPEM) throws Exception {
        String privateKeyPEMClean = privateKeyPEM
            .replace("-----BEGIN PRIVATE KEY-----", "")
            .replace("-----END PRIVATE KEY-----", "")
            .replaceAll("\\s", "");
        
        byte[] encoded = Base64.getDecoder().decode(privateKeyPEMClean);
        PKCS8EncodedKeySpec keySpec = new PKCS8EncodedKeySpec(encoded);
        KeyFactory kf = KeyFactory.getInstance("RSA");
        return kf.generatePrivate(keySpec);
    }

    public int getMaxCustomerId() throws Exception {
        String jdbcUrl = config.getSnowflakeUrl().replace(":443", "").replace("https://", "jdbc:snowflake://");
        
        Properties jdbcProps = new Properties();
        jdbcProps.put("account", config.getSnowflakeAccount());
        jdbcProps.put("user", config.getSnowflakeUser());
        jdbcProps.put("role", config.getRole());
        jdbcProps.put("warehouse", config.getWarehouse());
        jdbcProps.put("db", config.getDatabase());
        jdbcProps.put("schema", config.getSchema());
        
        PrivateKey privateKey = parsePrivateKey(config.getPrivateKey());
        jdbcProps.put("privateKey", privateKey);
        
        int maxId = 1;
        try (Connection conn = DriverManager.getConnection(jdbcUrl, jdbcProps);
             Statement stmt = conn.createStatement();
             ResultSet rs = stmt.executeQuery("SELECT MAX(CUSTOMER_ID) as MAX_ID FROM " + 
                                            config.getDatabase() + ".RAW.CUSTOMERS")) {
            if (rs.next()) {
                maxId = rs.getInt("MAX_ID");
            }
            logger.info("Max customer ID in database: {}", maxId);
        }
        
        return maxId;
    }

    public String getCustomerSegment(int customerId) {
        // For simplicity during high-volume streaming, randomly assign segments
        // to avoid frequent DB queries that could slow down ingestion
        String[] segments = {"Premium", "Standard", "Basic"};
        return segments[new java.util.Random().nextInt(segments.length)];
    }

    public void insertOrder(Order order) throws Exception {
        Map<String, Object> row = order.toMap();
        String offsetToken = "order_" + order.getOrderId();
        
        ordersChannel.appendRow(row, offsetToken);
        logger.debug("Order {} inserted with offset {}", order.getOrderId(), offsetToken);
    }

    public void insertOrders(List<Order> orders) throws Exception {
        if (orders.isEmpty()) {
            return;
        }
        
        String startOffset = "order_" + orders.getFirst().getOrderId();
        String endOffset = "order_" + orders.getLast().getOrderId();
        
        List<Map<String, Object>> rows = orders.stream()
                .map(Order::toMap)
                .collect(Collectors.toList());
        
        insertWithBackpressureRetry(ordersChannel, rows, startOffset, endOffset, "orders");
        this.lastOrdersOffset = endOffset;
        logger.debug("Inserted {} orders (offset range: {} to {})", 
                    orders.size(), startOffset, endOffset);
    }

    public void insertOrderItems(List<OrderItem> items) throws Exception {
        if (items.isEmpty()) {
            return;
        }
        
        String startOffset = "item_" + items.getFirst().getOrderItemId();
        String endOffset = "item_" + items.getLast().getOrderItemId();
        
        List<Map<String, Object>> rows = items.stream()
                .map(OrderItem::toMap)
                .collect(Collectors.toList());
        
        insertWithBackpressureRetry(orderItemsChannel, rows, startOffset, endOffset, "order_items");
        this.lastOrderItemsOffset = endOffset;
        logger.debug("Inserted {} order items (offset range: {} to {})", 
                    items.size(), startOffset, endOffset);
    }

    /**
     * Insert rows with exponential backoff retry for ReceiverSaturated errors.
     * 
     * @param channel The streaming channel to insert into
     * @param rows Data rows to insert
     * @param startOffset Starting offset token
     * @param endOffset Ending offset token
     * @param dataType Description of data type (for logging)
     */
    private void insertWithBackpressureRetry(
            SnowflakeStreamingIngestChannel channel,
            List<Map<String, Object>> rows,
            String startOffset,
            String endOffset,
            String dataType) throws Exception {
        
        int maxRetries = 10;
        double delay = 1.0;  // Initial delay in seconds
        double maxDelay = 30.0;
        
        for (int attempt = 0; attempt < maxRetries; attempt++) {
            try {
                channel.appendRows(rows, startOffset, endOffset);
                
                if (attempt > 0) {
                    logger.info("Successfully inserted {} {} after {} attempts",
                               rows.size(), dataType, attempt + 1);
                }
                return;
                
            } catch (Exception e) {
                String errorMsg = e.getMessage();
                
                // Check for ReceiverSaturated (HTTP 429) backpressure errors
                // The Java SDK wraps the error — message may contain "ReceiverSaturated",
                // "429", or "running full" depending on the SDK version
                if (errorMsg.contains("ReceiverSaturated") || errorMsg.contains("429") 
                        || errorMsg.contains("running full") || errorMsg.contains("cannot be accepted")) {
                    if (attempt < maxRetries - 1) {
                        logger.warn("Backpressure detected for {} (attempt {}/{}): " +
                                   "Channel buffers full. Retrying in {:.1f}s...",
                                   dataType, attempt + 1, maxRetries, delay);
                        Thread.sleep((long) (delay * 1000));
                        delay = Math.min(delay * 2, maxDelay);  // Exponential backoff
                        continue;
                    } else {
                        logger.error("Failed to insert {} {} after {} attempts: " +
                                    "Channel buffers remain saturated",
                                    rows.size(), dataType, maxRetries);
                        throw e;
                    }
                } else {
                    // Non-backpressure error, re-throw immediately
                    logger.error("Unexpected error inserting {}: {}", dataType, errorMsg);
                    throw e;
                }
            }
        }
    }

    public String getLatestOrderOffset() {
        return ordersChannel.getLatestCommittedOffsetToken();
    }

    public String getLatestOrderItemOffset() {
        return orderItemsChannel.getLatestCommittedOffsetToken();
    }

    /**
     * Wait until both channels have committed all in-flight data.
     * Polls each channel's latest committed offset token until it matches
     * the last offset we sent, confirming Snowflake has received everything.
     *
     * @param timeoutSeconds Maximum time to wait
     * @param pollIntervalMs Polling interval in milliseconds
     * @return true if both channels flushed within the timeout
     */
    public boolean waitForFlush(int timeoutSeconds, long pollIntervalMs) throws InterruptedException {
        logger.info("Waiting for all channels to flush in-flight data...");
        long start = System.currentTimeMillis();
        long timeoutMs = timeoutSeconds * 1000L;

        boolean ordersFlushed = (lastOrdersOffset == null);
        boolean itemsFlushed = (lastOrderItemsOffset == null);

        if (ordersFlushed && itemsFlushed) {
            logger.info("No data was sent — nothing to flush");
            return true;
        }

        while (System.currentTimeMillis() - start < timeoutMs) {
            if (!ordersFlushed) {
                String committed = ordersChannel.getLatestCommittedOffsetToken();
                if (lastOrdersOffset.equals(committed)) {
                    long elapsed = System.currentTimeMillis() - start;
                    logger.info("Channel orders flushed (offset {}) in {}ms", committed, elapsed);
                    ordersFlushed = true;
                }
            }

            if (!itemsFlushed) {
                String committed = orderItemsChannel.getLatestCommittedOffsetToken();
                if (lastOrderItemsOffset.equals(committed)) {
                    long elapsed = System.currentTimeMillis() - start;
                    logger.info("Channel order_items flushed (offset {}) in {}ms", committed, elapsed);
                    itemsFlushed = true;
                }
            }

            if (ordersFlushed && itemsFlushed) {
                long total = System.currentTimeMillis() - start;
                logger.info("All channels flushed in {}ms", total);
                return true;
            }

            Thread.sleep(pollIntervalMs);
        }

        // Timeout — log which channels are still pending
        if (!ordersFlushed) {
            String committed = ordersChannel.getLatestCommittedOffsetToken();
            logger.warn("Channel orders NOT flushed after {}s (expected: {}, committed: {})",
                        timeoutSeconds, lastOrdersOffset, committed);
        }
        if (!itemsFlushed) {
            String committed = orderItemsChannel.getLatestCommittedOffsetToken();
            logger.warn("Channel order_items NOT flushed after {}s (expected: {}, committed: {})",
                        timeoutSeconds, lastOrderItemsOffset, committed);
        }
        return false;
    }

    public boolean waitForFlush() throws InterruptedException {
        return waitForFlush(120, 2000);
    }

    public void close() {
        try {
            logger.info("Closing channels...");
            if (ordersChannel != null) ordersChannel.close();
            if (orderItemsChannel != null) orderItemsChannel.close();
            
            logger.info("Closing clients...");
            if (ordersClient != null) ordersClient.close();
            if (orderItemsClient != null) orderItemsClient.close();
            
            logger.info("Snowpipe Streaming manager closed successfully");
        } catch (Exception e) {
            logger.error("Error closing Snowpipe Streaming manager", e);
        }
    }
}
